### PR TITLE
Unwatch users when updates are no longer needed

### PR DESCRIPTION
**Pull request by mathiascode**
Currently, the list of watched users grows indefinitely until we disconnect/quit. Unwatch users when we no longer need to track them.

### CODE REVIEW

File: pynicotine/interests.py [hunk 1]
```diff
@@ -134,6 +144,15 @@ def request_similar_users(self):
...
         """Server code 110 and 112."""
 
+        for username in self.similar_users:
+            if username not in msg.users:
+                core.users.unwatch_user(username, mode="interests")
+
+        self.similar_users.clear()
+
         for username in msg.users:
+            rating = msg.users[username] if isinstance(msg.users, dict) else None
```
Comment (slook) [inline pynicotine/interests.py:154]: `msg.users` should always be a `dict` even if it is empty, so `isinstance()` shouldn't be required.

File: pynicotine/users.py [hunk 2]
```diff
@@ -28,14 +28,16 @@
...
-    __slots__ = ("username", "upload_speed", "files", "folders")
+    __slots__ = ("username", "upload_speed", "files", "folders", "modes", "implicit")
 
     def __init__(self, username):
 
         self.username = username
         self.upload_speed = None
         self.files = None
         self.folders = None
+        self.modes = set()
```
Comment (slook) [inline pynicotine/users.py:39]: String matching to identify different modes isn't very nice.

Consider defining a new class such as `WatchMode()` for this purpose.

File: pynicotine/users.py [hunk 3]
```diff
@@ -28,14 +28,16 @@
...
+        self.implicit = True  # True if no WatchUser message was sent (i.e. receiving updates for room user)
```
Comment (slook) [inline pynicotine/users.py:40]: Consider a `set()` to contain all the `rooms` that a user is in, avoids doing string concatenations.

If the set is empty the user would not be in any of the rooms we are in.

File: pynicotine/buddies.py [hunk 4]
```diff
@@ -165,7 +165,7 @@ def add_buddy(self, username):
         events.emit("add-buddy", username, user_data)
 
         # Request user status, speed and number of shared files
-        core.users.watch_user(username)
+        core.users.watch_user(username, mode="userlist")
```
Comment (slook) [inline pynicotine/buddies.py:168]: Ambiguity with normal user. "userlist" -> "buddy"

**Review by slook — CHANGES_REQUESTED** [review]

File: pynicotine/interests.py (hunk 1 above)
Comment (mathiascode) [reply pynicotine/interests.py:154]: It can be a list or dict depending on which server message it is (one doesn't include ratings). I'll probably come up with something better.

File: pynicotine/users.py (hunk 2 above)
Comment (mathiascode) [reply pynicotine/users.py:39]: It doesn't feel right to make the users module responsible for knowing every mode in the application. Ideally you could call watch_user() from anywhere (say, a plugin) without being required to add a new mode here.

File: pynicotine/users.py (hunk 3 above)
Comment (mathiascode) [reply pynicotine/users.py:40]: Similar concerns as the previous comment. I don't think the users module should know this.

Comment (mathiascode) [reply pynicotine/users.py:40]: I'll get rid of `mode.startswith("chatroom_")` in favor of a boolean parameter for watch_user().

File: pynicotine/users.py (hunk 2 above)
Comment (slook) [reply pynicotine/users.py:39]: Perhaps get a value without needing to pass an argument with something in https://docs.python.org/3/library/inspect.html#module-inspect

File: pynicotine/buddies.py (hunk 4 above)
Comment (slook) [reply pynicotine/buddies.py:168]: Rather than "buddy", actually "buddies" would match what is used in `core.enabled_components`.

**Comment by mathiascode** [comment]
I renamed mode -> context, seemed a little clearer to me.

Also refactored transfer user watching so it doesn't unwatch/rewatch when changing the state of the last remaining transfer.

File: pynicotine/users.py [hunk 5]
```diff
@@ -116,20 +118,86 @@ def request_set_status(self, status):
...
             return
 
-        if username in self.watched:
+        watched_user = self.watched.get(username)
+
+        if watched_user is None:
+            self.watched[username] = watched_user = WatchedUser(username)
+
+        if not context:
+            context = "unknown"
```
Comment (slook) [inline pynicotine/users.py:142]: Why is this needed?

If the function call is made improperly, then an error message should be logged.

File: pynicotine/users.py [hunk 6]
```diff
@@ -116,20 +118,86 @@ def request_set_status(self, status):
...
+            # We can't unwatch ourselves
+            return
+
+        watched_user = self.watched.get(username)
+
+        if watched_user is None:
+            return
+
+        if not context:
+            context = "unknown"
```
Comment (slook) [inline pynicotine/users.py:176]: If the context is unspecified, then perhaps all contexts ought to be discarded?

File: pynicotine/users.py (hunk 5 above)
Comment (mathiascode) [reply pynicotine/users.py:142]: I'd like to get the change into 3.3.x, but wasn't sure if any plugins are calling this function. I'll just mark watch_user() with no context as deprecated, and make the context mandatory in 3.4.0.

File: pynicotine/users.py (hunk 6 above)
Comment (mathiascode) [reply pynicotine/users.py:176]: Too easy to misuse and break things. I'll make the context mandatory, since it's a new function.
